Lib/multiprocessing/resource_tracker.py

```diff
@@ -81,6 +83,8 @@ def ensure_running(self):
         This can be run from any process.  Usually a child process will use
         the resource created by its parent.'''
         with self._lock:
+            if self._lock._recursion_count() > 1:
+                return
```

Comment:
Yeah, I will add a comment explaining it.

Yes, on a reentrant call, it will not do what it says. I have no idea how to avoid that. Reentrancy is much harder than multi-threading: you cannot wait for the "other thread", since it's waiting for you :-)

The real solution would be to avoid reentrant calls from finalizers, but that will be harder to solve and not backportable.